tensorflow/core/kernels/dml_batch_to_space_op.cc

```diff
@@ -316,96 +353,165 @@ class DmlBatchToSpaceKernel : public DmlKernel {
     absl::Span<const int64> internal_block_sizes =
         init_helper->GetInternalBlockSizes();
 
-    uint32_t batch_size = internal_input_shape.dim_size(0);
-    uint32_t block_shape_product = std::accumulate(internal_block_sizes.begin(),
-                                                   internal_block_sizes.end(),
-                                                   1, std::multiplies<int64>());
+    dml::Expression output_before_slice;
 
-    // Reshape the input into [block_shape[0], ..., block_shape[M-1], batch /
-    // prod(block_shape), input_shape[1], ..., input_shape[N-1]]
-    dml::TensorDesc::Dimensions reshaped_sizes;
-    for (int i = 0; i < internal_block_sizes.size(); ++i) {
-      reshaped_sizes.push_back(internal_block_sizes[i]);
-    }
+    // For 4d cases we can use the DML Depth to Space operator
+    if (internal_input_shape.dims() == 4) {
+      auto input_sizes = input_desc.desc.GetSizes();
+      auto input_strides = input_desc.desc.GetStrides();
 
-    reshaped_sizes.push_back(internal_input_shape.dim_size(0) /
-                             block_shape_product);
+      if (input_strides.empty())
+      {
+        input_strides = ComputePackedStrides(input_sizes);
+      }
 
-    for (int i = 1; i < internal_input_shape.dims(); ++i) {
-      reshaped_sizes.push_back(internal_input_shape.dim_size(i));
-    }
+      // Permute input NCHW->WNCH
+      dml::TensorDimensions input_tensor_sizes = {
+          input_sizes[3], input_sizes[0], input_sizes[1], input_sizes[2]};
+      dml::TensorStrides input_tensor_strides = {
+          input_strides[3], input_strides[0], input_strides[1],
+          input_strides[2]};
```

Comment:
We already have a way of doing that, which is a bit less error-prone (e.g. look at dml_conv_ops.cc):

```
    // The dimensions of the filter tensor are laid out a little differently
    // than what DML expects
    auto filter_layout = {H, W, C, N};

    // The layout of the input/output tensors is determined by the "data_format"
    // attribute
    auto input_output_layout =
        GetDmlTensorLayout(init_helper->GetDataFormat(), kNchwDimensionCount);
```